helix-view/src/document.rs

```diff
@@ -393,57 +393,64 @@ pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
     encoding: &'static encoding::Encoding,
     rope: &'a Rope,
 ) -> Result<(), Error> {
-    // Text inside a `Rope` is stored as non-contiguous blocks of data called
-    // chunks. The absolute size of each chunk is unknown, thus it is impossible
-    // to predict the end of the chunk iterator ahead of time. Instead, it is
-    // determined by filtering the iterator to remove all empty chunks and then
-    // appending an empty chunk to it. This is valuable for detecting when all
-    // chunks in the `Rope` have been iterated over in the subsequent loop.
-    let iter = rope
-        .chunks()
-        .filter(|c| !c.is_empty())
-        .chain(std::iter::once(""));
-    let mut buf = [0u8; BUF_SIZE];
-    let mut encoder = encoding.new_encoder();
-    let mut total_written = 0usize;
-    for chunk in iter {
-        let is_empty = chunk.is_empty();
-        let mut total_read = 0usize;
-
-        // An inner loop is necessary as it is possible that the input buffer
-        // may not be completely encoded on the first `encode_from_utf8()` call
-        // which would happen in cases where the output buffer is filled to
-        // capacity.
-        loop {
-            let (result, read, written, ..) =
-                encoder.encode_from_utf8(&chunk[total_read..], &mut buf[total_written..], is_empty);
+    if encoding == encoding::UTF_16BE {
+        for ch in rope.chunks().flat_map(|chunk| chunk.encode_utf16()) {
+            writer.write_u16(ch).await?;
+        }
+    } else if encoding == encoding::UTF_16LE {
+        for ch in rope.chunks().flat_map(|chunk| chunk.encode_utf16()) {
+            writer.write_u16_le(ch).await?;
```

Comment:
Calling the writer directly is pretty ineffient and will cause a ton of syscalls. You want to instead use the buffering code below.

I think a better approach would be to essentially create an.

```
enum Encoder {
 Utf16Be
 Utf16Le,
 encodingRs(..)
}
```
and then implement utf16 encoding with the same interface the encoding_rs uses so we don't duplicate the buffering logic here